File: datasets/C++/Deepseek/Algorithm/medium5.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<string> determine_winner(const vector<pair<int, vector<pair<int, int>>>>& games) {
    vector<string> results;
    for (const auto& game : games) {
        int n = game.first;
        const auto& trees = game.second;
        int total_xor = 0;
        for (const auto& tree : trees) {
            int m = tree.first;
            int k = tree.second;
            if (m == 1) {
                total_xor ^= 1;
            } else {
                if ((m - 1) % (k + 1) == 0) {
                    // grundy number is 0, do nothing
                } else {
                    total_xor ^= 1;
                }
            }
        }
        if (total_xor != 0) {
            results.push_back("BOB");
        } else {
            results.push_back("BEN");
        }
    }
    return results;
}

int main() {
    int g;
    cin >> g;
    vector<pair<int, vector<pair<int, int>>>> games;
    for (int i = 0; i < g; ++i) {
        int n;
        cin >> n;
        vector<pair<int, int>> trees;
        for (int j = 0; j < n; ++j) {
            int m, k;
            cin >> m >> k;
            trees.emplace_back(m, k);
        }
        games.emplace_back(n, trees);
    }
    vector<string> results = determine_winner(games);
    for (const auto& result : results) {
        cout << result << endl;
    }
    return 0;
}

File: datasets/Python/Deepseek/Algorithm/medium5.py
import sys

def determine_winner(games):
    results = []
    for game in games:
        n = game[0]
        trees = game[1]
        total_xor = 0
        for m, k in trees:
            if m == 1:
                total_xor ^= 1
            else:
                # The pattern observed is that the grundy number is 0 if (m - 1) mod (k + 1) == 0
                # Otherwise, it's 1 or something else. Need to verify.
                # For the problem, perhaps the grundy number is (m - 1) % (k + 1) == 0 ? 0 : 1?
                # But sample inputs may suggest otherwise. Need more analysis.
                # Let's assume for now that grundy is 0 if m is 1 mod (k+1), else 1.
                if (m - 1) % (k + 1) == 0:
                    pass  # grundy 0, no xor
                else:
                    total_xor ^= 1
        if total_xor != 0:
            results.append("BOB")
        else:
            results.append("BEN")
    return results

def main():
    input_lines = sys.stdin.read().split()
    ptr = 0
    g = int(input_lines[ptr])
    ptr += 1
    games = []
    for _ in range(g):
        n = int(input_lines[ptr])
        ptr += 1
        trees = []
        for __ in range(n):
            m, k = map(int, input_lines[ptr:ptr+2])
            ptr += 2
            trees.append((m, k))
        games.append((n, trees))
    results = determine_winner(games)
    print('\n'.join(results))

if __name__ == "__main__":
    main()